c_glib/arrow-glib/gobject-type.h

```diff
@@ -0,0 +1,97 @@
+                                                                        \
+  GType object_name ## _get_type(void) G_GNUC_CONST;                    \
+                                                                        \
+  static inline ObjectName *                                            \
+  MODULE_NAME ## _ ## OBJECT_NAME(gpointer object)                      \
+  {                                                                     \
+    return G_TYPE_CHECK_INSTANCE_CAST(object,                           \
+                                      object_name ## _get_type(),       \
+                                      ObjectName);                      \
+  }                                                                     \
+                                                                        \
+  static inline ObjectName ## Class *                                   \
+  MODULE_NAME ## _ ## OBJECT_NAME ## _CLASS(gpointer klass)             \
+  {                                                                     \
+    return G_TYPE_CHECK_CLASS_CAST(klass,                               \
+                                   object_name ## _get_type(),          \
+                                   ObjectName ## Class);                \
+  }                                                                     \
+                                                                        \
+  static inline gboolean                                                \
+  MODULE_NAME ## _IS_ ## OBJECT_NAME(gpointer object)                   \
+  {                                                                     \
+    return G_TYPE_CHECK_INSTANCE_TYPE(object,                           \
+                                      object_name ## _get_type());      \
+  }                                                                     \
+                                                                        \
+  static inline gboolean                                                \
+  MODULE_NAME ## _IS_ ## OBJECT_NAME ## _CLASS(gpointer klass)          \
+  {                                                                     \
+    return G_TYPE_CHECK_CLASS_TYPE(klass,                               \
+                                   object_name ## _get_type());         \
+  }                                                                     \
+                                                                        \
+  static inline ObjectName ## Class *                                   \
+  MODULE_NAME ## _ ## ObjectName ## _GET_CLASS(gpointer object)         \
+  {                                                                     \
+    return G_TYPE_INSTANCE_GET_CLASS(object,                            \
+                                     object_name ## _get_type(),        \
+                                     ObjectName ## Class);              \
+  }
+#endif
```

Comment:
Since GLib is LGPL (which is Category X) I wanted to check that this code did not come from that codebase

Reply:
GLib's implementation is here: https://git.gnome.org/browse/glib/tree/gobject/gtype.h#n1395

This implementation generates codes based on https://github.com/apache/arrow/blob/c1492bf4a01600c838c59ebef7dd88694cca41c3/c_glib/arrow-glib/array.h#L28-L60

Defining macros such as `GARROW_ARRAY()` by "static inline" function idea is taken from GLib's implementation. Others aren't taken.

It seems that the idea (implementing convenient functions by "static inline" function instead of macro) is a generic C technique. I know that Ruby is also uses the technique such as https://github.com/ruby/ruby/blob/trunk/include/ruby/ruby.h#L307-L320  .

So I think that this code isn't LGPL-ed work.

Reply:
No problem, thank you. Since this was feature backporting issue I just wanted to check. Reusing implementation ideas is no problem so long as actual lines of Category X code are not being redistributed or otherwise directly ported

Reply:
Thanks for checking license issue!